src/peft/tuners/cartridge/model.py

```diff
@@ -0,0 +1,116 @@
+# Copyright 2025-present the HuggingFace Inc. team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+import torch
+
+
+class CartridgeEncoder(torch.nn.Module):
+    """
+    A parameterized prefix KV cache.
+
+    The parameters are stored in the same flattened layout as `PrefixEncoder` output: `[num_virtual_tokens, num_layers
+    * 2 * token_dim]`, where `token_dim` is per-head hidden size times number of heads (after any GQA adjustment
+    performed by `_prepare_prompt_learning_config`).
+
+    If `num_frozen_tokens > 0`, the first `num_frozen_tokens` virtual tokens are stored as a non-trainable parameter,
+    and the remaining tokens are trainable.
+    """
+
+    def __init__(self, config):
+        super().__init__()
+        self.config = config
+
+        num_virtual_tokens = config.num_virtual_tokens
+        hidden = config.num_layers * 2 * config.token_dim
+        num_frozen_tokens = int(getattr(config, "num_frozen_tokens", 0) or 0)
+        if num_frozen_tokens < 0 or num_frozen_tokens > num_virtual_tokens:
```

Comment:
This could already be checked in the `__post_init__` of the config, right? The earlier we error, the better.